This is updated patch that fixes interface renaming broken by
systemd commit 97595710b77aa162ca5e20da57d0a1ed7355eaad.

udev: network device renaming - immediately give up if the target name isn't available

We want to to keep supporting systems with 75-persistent-net-generator.rules
generated names. Re-apply this old hack to make the renaming less likely to fail.

--- systemd-stable-246.1/src/udev/udev-event.c.orig	2020-08-07 17:09:53.000000000 +0200
+++ systemd-stable-246.1/src/udev/udev-event.c	2020-08-15 11:04:10.263428290 +0200
@@ -820,6 +820,8 @@
         sd_device *dev;
         unsigned flags;
         int ifindex, r;
+        char name[IFNAMSIZ];
+        int loop;
 
         assert(event);
 
@@ -861,6 +863,32 @@
                 return log_device_debug_errno(event->dev_db_clone, r, "Failed to update database under /run/udev/data/: %m");
 
         r = rtnl_set_link_name(&event->rtnl, ifindex, event->name);
+        /* keep trying if the destination interface name already exists */
+        if (r != -EEXIST)
+                goto out;
+
+        /* free our own name, another process may wait for us */
+        snprintf(name, IFNAMSIZ, "rename%u", ifindex);
+        r = rtnl_set_link_name(&event->rtnl, ifindex, name);
+        if (r < 0)
+                goto out;
+
+        /* log temporary name */
+        log_device_debug(dev, "Network interface %i is renamed from '%s' to '%s'", ifindex, oldname, event->name);
+
+        /* wait a maximum of 90 seconds for our target to become available */
+        strscpy(name, IFNAMSIZ, event->name);
+        loop = 90 * 20;
+        while (loop--) {
+                const struct timespec duration = { 0, 1000 * 1000 * 1000 / 20 };
+
+                nanosleep(&duration, NULL);
+
+                r = rtnl_set_link_name(&event->rtnl, ifindex, name);
+                if (r != -EEXIST)
+                        goto out;
+        }
+out:
         if (r < 0)
                 return log_device_error_errno(dev, r, "Failed to rename network interface %i from '%s' to '%s': %m",
                                               ifindex, oldname, event->name);
